server/src/main/java/org/elasticsearch/search/SearchService.java

```diff
@@ -363,37 +371,66 @@ private void loadOrExecuteQueryPhase(final ShardSearchRequest request, final Sea
         }
     }
 
-    public void executeQueryPhase(ShardSearchRequest request, SearchShardTask task, ActionListener<SearchPhaseResult> listener) {
+    public void executeQueryPhase(ShardSearchRequest request, boolean keepStatesInContext,
+                                  SearchShardTask task, ActionListener<SearchPhaseResult> listener) {
         assert request.canReturnNullResponseIfMatchNoDocs() == false || request.numberOfShards() > 1
             : "empty responses require more than one shard";
-        IndexService indexService = indicesService.indexServiceSafe(request.shardId().getIndex());
-        IndexShard shard = indexService.getShard(request.shardId().id());
+        final IndexShard shard = getShard(request);
         rewriteAndFetchShardRequest(shard, request, new ActionListener<ShardSearchRequest>() {
             @Override
             public void onResponse(ShardSearchRequest orig) {
+                final ReaderContext readerContext = createOrGetReaderContext(orig, keepStatesInContext);
+                final Releasable markAsUsed = readerContext.markAsUsed();
                 if (orig.canReturnNullResponseIfMatchNoDocs()) {
+                    assert orig.scroll() == null;
                     // we clone the shard request and perform a quick rewrite using a lightweight
                     // searcher since we are outside of the search thread pool.
                     // If the request rewrites to "match none" we can shortcut the query phase
                     // entirely. Otherwise we fork the execution in the search thread pool.
                     ShardSearchRequest canMatchRequest = new ShardSearchRequest(orig);
-                    try (Engine.Searcher searcher = shard.acquireCanMatchSearcher()) {
-                        QueryShardContext context = indexService.newQueryShardContext(canMatchRequest.shardId().id(), searcher,
-                            canMatchRequest::nowInMillis, canMatchRequest.getClusterAlias());
+                    try (Engine.Searcher searcher = readerContext.acquireSearcher(Engine.CAN_MATCH_SEARCH_SOURCE)) {
+                        QueryShardContext context = readerContext.indexService().newQueryShardContext(canMatchRequest.shardId().id(),
+                            searcher, canMatchRequest::nowInMillis, canMatchRequest.getClusterAlias());
                         Rewriteable.rewrite(canMatchRequest.getRewriteable(), context, true);
-                    } catch (Exception exc) {
-                        listener.onFailure(exc);
-                        return;
+                    } catch (IOException exc) {
+                        try (markAsUsed) {
+                            listener.onFailure(exc);
+                        } finally {
+                            processFailure(request, readerContext, exc);
+                        }
```

Comment:
@jimczi I think we should terminate here after handling an IOException?

Reply:
Good catch, I pushed https://github.com/elastic/elasticsearch/pull/56480/commits/05dd518ed1994848e43e126434ba5434a9277294